src/modules/simulator/simulator_mavlink.cpp

```diff
@@ -589,6 +589,36 @@ void *Simulator::sending_trampoline(void * /*unused*/)
 	return nullptr;
 }
 
+void Simulator::send_sim_tone()
+{
+	uint64_t time_now = hrt_absolute_time();
+
+	// Don't grab the next tone until the current one is scheduled to be done.
+	if (_tone_timer < time_now) {
+
+		struct {
+			unsigned int frequency = 0;
+			unsigned int duration = 0;
+			unsigned int silence = 0;
+			uint8_t volume = 0;
+		} tone_info = {};
+
+		if (_tunes.get_next_note(tone_info.frequency, tone_info.duration, tone_info.silence) > 0) {
+
+			_tone_timer = time_now + tone_info.duration + tone_info.silence;
+
+			if (tone_info.frequency > 0 && tone_info.duration > 0 && tone_info.volume > 0) {
+				mavlink_play_tune_t tune_msg {};
+				mavlink_message_t message {};
+
+				memcpy(tune_msg.tune, &tone_info, sizeof(tone_info));
```

Comment:
Why are you creating the tone_info struct and then copying from it into the mavlink_play_tune_t struct? Couldn't you use the mavlink_play_tune_t struct from the beginning. Also, I wonder if the padding could be different if mavlink_play_tune_t is packed and the struct above is not.

Reply:
Thanks @julianoes , I think the reason this was done originally is that the two two structs differ in how the `get_next_note()` method operates.  `get_next_note()` gives back `unsigned int` values for frequency, duration, silence, and volume, and that 16 bytes of data needs to get packed into the `mavlink_play_tune.tune[]` 30 byte char array.  Is there a better way I could try to implement this?  Thanks!

Reply:
Aha, now I understand. Does the `tone_info` struct need to be packed then? Otherwise padding might mess with the format, right?

Reply:
I think you're right about that, but I'm missing something there.  When I try to pack that struct it can't bind the values in the `get_next_note()` call.

Reply:
Hm, I see. Where is the counterpart for this function? The one sending the tunes?

Reply:
Or asking differently? How can I test/use this?

Reply:
FYI @JonasVautherin

Reply:
Hi @julianoes , the counterpart to this PR is [here](https://github.com/PX4/jMAVSim/pull/92), now closed.  I've tried again to resolve conflicts and so far had only concluded that I don't actually understand everything going on. :/  Give me a little bit of time to try to figure everything out, otherwise, perhaps, we can just close out this PR.